File: 907-koko-eating-bananas/koko-eating-bananas.cpp
#include <cmath>
#include <iostream>
#include <vector>
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1, r = *max_element(piles.begin(),piles.end());
        
        while(l<=r){
            int k = r+(l-r)/2;
            long long time = 0;
            for(int i = 0; i< piles.size();i++){
                time +=ceil(static_cast<double>(piles[i])/k);
            }
            if(time>h){
                l = k+1;
            }
            else{
                r = k-1;
            }
        }
        return l;
    }
};

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l,r = 1,max(piles)
        ret = r
        while (l<=r):
            k = (r+l)//2
            tot = 0
            for p in piles:
                tot+= math.ceil(p/k)
            if tot >h:
                l = k+1
            else:
                ret = k
                r = k-1
        return ret